### ISSUE TITLE
Multiprocessing map gets stuck if doing inference on loaded model

### ISSUE THREAD

**Issue by GPaolo** [pytorch/pytorch#35472]
## 🐛 Bug

When doing inference on a loaded model through the torch.multiprocessing.map function the code gets stuck. The same does not apply if I use a model that is not loaded (e.g. I just instantiate one with random weights) or if I do not use multiprocessing but use the loaded model. 

## To Reproduce

A simple code snippet to show the structure of my code is:

```python
# This is evaluator.py
class Evaluator(nn.Module):
  def __init__(self):
    self.model = nn.Linear(10, 5, bias=False)
    ckpt = torch.load(filepath, map_location=self.device)
    self.load_state_dict(ckpt)

  def forward(self, x):
    return self.model(x)
```

```python
# This is the main file main.py

import torch.multiprocessing as mp
from evaluator import Evaluator
pool = None

class Main():
  def __init__(self):
    global pool
    eval = Evaluator()
    pool = mp.Pool(processes=4)

  def execute(self, data):
    global pool
    results = pool.map(eval.forward, data)
    return results
```
```python 
# This is the script
from main import Main

if  __name__ == "__main__":
  main = Main()
  data = [list of tensors]
  results = main.execute(data)
```

## Expected behavior

What I expect is that both when I load the model or when I don't load the model the map function would return. What happens instead is that if I load the model the map function remains stuck.

This does not happens if I do not use the map, but a normal for loop, without any multiprocess.

## Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] torch==1.4.0
[pip3] torchvision==0.5.0
[conda] Could not collect

**Comment by mruberry** [comment]
Hey @GPaolo, sorry to hear you're having a problem. Multiprocessing with deep learning frameworks can be a little tricky, and hanging is never fun. What happens if you load the model in each process? Also, we have some articles about multiprocessing with PyTorch that may be helpful:

https://pytorch.org/docs/stable/multiprocessing.html
https://pytorch.org/docs/stable/notes/multiprocessing.html

**Comment by GPaolo** [comment]
Hi, thanks for the reply!
If I load the model once per process it works without problems. 
The hanging only happens if the model is loaded externally and then passed to the processes. What is weird is that the hanging does not happen if I create the model externally (without loading one) and then pass it to the processes. 

I have read the articles. And actually have a question about them, the map and apply functions are not mentioned. Does it mean that they do not work with torch models or do they still work as in the multiprocessing module and for this reason they are not mentioned?

**Comment by mruberry** [comment]
Deep learning frameworks like PyTorch often have particular multiprocessing patterns to handle context initialization and cloning. The hanging behavior you're describing is, unfortunately, pretty easy to hit, and it's too bad it doesn't give a more informative error message. I actually included a section on this in a PyTorch/XLA tutorial I wrote, see: 

https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/multi-core-alexnet-fashion-mnist.ipynb

I would stick with the recommended patterns for multiprocessing as a result. Other things may work but aren't guaranteed to do so.

**Comment by sheep94lion** [comment]
> ##  Bug
> When doing inference on a loaded model through the torch.multiprocessing.map function the code gets stuck. The same does not apply if I use a model that is not loaded (e.g. I just instantiate one with random weights) or if I do not use multiprocessing but use the loaded model.
> 
> I guess is somewhat related to: [this issue 16943](https://github.com/pytorch/pytorch/issues/16943)
> 
> ## To Reproduce
> A simple code snippet to show the structure of my code is:
> 
> ```python
> # This is evaluator.py
> class Evaluator(nn.Module):
>   def __init__(self):
>     self.model = nn.Linear(10, 5, bias=False)
>     ckpt = torch.load(filepath, map_location=self.device)
>     self.load_state_dict(ckpt)
> 
>   def forward(self, x):
>     return self.model(x)
> ```
> 
> ```python
> # This is the main file main.py
> 
> import torch.multiprocessing as mp
> from evaluator import Evaluator
> pool = None
> 
> class Main():
>   def __init__(self):
>     global pool
>     eval = Evaluator()
>     pool = mp.Pool(processes=4)
> 
>   def execute(self, data):
>     global pool
>     results = pool.map(eval.forward, data)
>     return results
> ```
> 
> ```python
> # This is the script
> from main import Main
> 
> if  __name__ == "__main__":
>   main = Main()
>   data = [list of tensors]
>   results = main.execute(data)
> ```
> 
> ## Expected behavior
> What I expect is that both when I load the model or when I don't load the model the map function would return. What happens instead is that if I load the model the map function remains stuck.
> 
> This does not happens if I do not use the map, but a normal for loop, without any multiprocess.
> 
> ## Environment
> PyTorch version: 1.4.0
> Is debug build: No
> CUDA used to build PyTorch: 10.1
> 
> OS: Ubuntu 16.04.6 LTS
> GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
> CMake version: version 3.5.1
> 
> Python version: 3.7
> Is CUDA available: No
> CUDA runtime version: No CUDA
> GPU models and configuration: No CUDA
> Nvidia driver version: No CUDA
> cuDNN version: No CUDA
> 
> Versions of relevant libraries:
> [pip3] numpy==1.18.1
> [pip3] torch==1.4.0
> [pip3] torchvision==0.5.0
> [conda] Could not collect

Will this be faster than infer each element of data sequentially?

**Comment by mruberry** [comment]
@sheep94lion Do you mean is multiprocessing faster than running in a single process? The answer depends on your program and your hardware. If you can distribute your computation effectively and you have multiple accelerators then it's likely to give you a significant speedup.

**Comment by sheep94lion** [comment]
> @sheep94lion Do you mean is multiprocessing faster than running in a single process? The answer depends on your program and your hardware. If you can distribute your computation effectively and you have multiple accelerators then it's likely to give you a significant speedup.

Thanks for your reply. If I only have one GPU and say 10000 images to process (inference), will multiprocess give me any speedup? Or should I use a properly large batch_size (instead of processing the image one by one) to reduce the overall processing time?

**Comment by mruberry** [comment]
The answer to your question is too dependent on your specific problem. You could try both.

**Comment by IvyGongoogle** [comment]
@GPaolo  I meet the similar question. I use cpu,  and get struck in function `load_state_dict`.  Can you give some advises?  Thanks.

**Comment by mruberry** [comment]
@IvyGongoogle You might want to try asking your question on the PyTorch forums at https://discuss.pytorch.org/.

**Comment by GPaolo** [comment]
> > ## Bug
> > When doing inference on a loaded model through the torch.multiprocessing.map function the code gets stuck. The same does not apply if I use a model that is not loaded (e.g. I just instantiate one with random weights) or if I do not use multiprocessing but use the loaded model.
> > I guess is somewhat related to: [this issue 16943](https://github.com/pytorch/pytorch/issues/16943)
> > ## To Reproduce
> > A simple code snippet to show the structure of my code is:
> > ```python
> > # This is evaluator.py
> > class Evaluator(nn.Module):
> >   def __init__(self):
> >     self.model = nn.Linear(10, 5, bias=False)
> >     ckpt = torch.load(filepath, map_location=self.device)
> >     self.load_state_dict(ckpt)
> > 
> >   def forward(self, x):
> >     return self.model(x)
> > ```
> > 
> > 
> > ```python
> > # This is the main file main.py
> > 
> > import torch.multiprocessing as mp
> > from evaluator import Evaluator
> > pool = None
> > 
> > class Main():
> >   def __init__(self):
> >     global pool
> >     eval = Evaluator()
> >     pool = mp.Pool(processes=4)
> > 
> >   def execute(self, data):
> >     global pool
> >     results = pool.map(eval.forward, data)
> >     return results
> > ```
> > 
> > 
> > ```python
> > # This is the script
> > from main import Main
> > 
> > if  __name__ == "__main__":
> >   main = Main()
> >   data = [list of tensors]
> >   results = main.execute(data)
> > ```
> > 
> > 
> > ## Expected behavior
> > What I expect is that both when I load the model or when I don't load the model the map function would return. What happens instead is that if I load the model the map function remains stuck.
> > This does not happens if I do not use the map, but a normal for loop, without any multiprocess.
> > ## Environment
> > PyTorch version: 1.4.0
> > Is debug build: No
> > CUDA used to build PyTorch: 10.1
> > OS: Ubuntu 16.04.6 LTS
> > GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
> > CMake version: version 3.5.1
> > Python version: 3.7
> > Is CUDA available: No
> > CUDA runtime version: No CUDA
> > GPU models and configuration: No CUDA
> > Nvidia driver version: No CUDA
> > cuDNN version: No CUDA
> > Versions of relevant libraries:
> > [pip3] numpy==1.18.1
> > [pip3] torch==1.4.0
> > [pip3] torchvision==0.5.0
> > [conda] Could not collect
> 
> Will this be faster than infer each element of data sequentially in the same process?

With this exact code yes, but this is just part of a bigger code that I have and for how the code is structured, having evaluation like this in parallel gives a significant speedup

**Comment by BlackDChase** [comment]
@GPaolo I had a similar issue, then I did what @mruberry suggested:
The first block of code works, the second block does not. (Both separate but similar files)
```py
import torch 
from torch import multiprocessing,nn 
 
output_size = 13 
input_dim = output_size 
hidden_dim = 128 
layer_dim = 1 
 
lstm = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
torch.save(lstm.state_dict(),"ENV_MODEL/plain.pt") 
 
def run(state): 
    model = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
    model.load_state_dict(torch.load("ENV_MODEL/plain.pt")) 
    x=model(state) 
    return x 
 
state=torch.rand(1,10,13) 
threads=[] 
for i in range(5): 
    p = multiprocessing.Process(target=run,args=(state,)) 
    threads.append(p) 
for i in threads: 
    i.start() 
for i in threads: 
    i.join() 
```
This is done just to experiment, but it got stuck. 
```py
import torch 
from torch import multiprocessing,nn 
 
output_size = 13 
input_dim = output_size 
hidden_dim = 128 
layer_dim = 1 
 
lstm = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
torch.save(lstm.state_dict(),"ENV_MODEL/plain.pt") 
 
def run(state): 
    model = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
    model.load_state_dict(torch.load("ENV_MODEL/plain.pt")) 
    x=model(state) 
    return x 
 
state=torch.rand(1,10,13) 
threads=[] 
for i in range(5): 
    p = multiprocessing.Process(target=run,args=(state,)) 
    threads.append(p) 
for i in threads: 
    i.start() 
for i in threads: 
    i.join() 
print("CHECKPOINT One") 
run(state) 
print("CHECKPOINT Two")
threads=[] 
for i in range(5): 
    p = multiprocessing.Process(target=run,args=(state,)) 
    threads.append(p) 
for i in threads: 
    i.start() 
for i in threads: 
    i.join()
print("CHECKPOINT Three")
```

My actual code does involve more wrappers and class structure but later would be what o would do in a zest. Any suggestion for the latter? (It got stuck after CHECKPOINT 2.)

**Comment by GPaolo** [comment]
Sorry I cannot help you. I ended up restructuring my code so I have a single model in the main process that is fed the data generated in the multiprocessing processes.

**Comment by BlackDChase** [comment]
@GPaolo No worries ... I restructured the code as well. It turns out, that if you load a model in parent and then load a similar/same model in a child it will get stuck. (Don't know why)
But if you load it first in the child process and then in the parent. It will work.  Refer this:
> 
> ```python
> import torch 
> from torch import multiprocessing,nn 
>  
> output_size = 13 
> input_dim = output_size 
> hidden_dim = 128 
> layer_dim = 1 
>  
> lstm = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
> torch.save(lstm.state_dict(),"ENV_MODEL/plain.pt") 
>  
> def run(state): 
>     model = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
>     model.load_state_dict(torch.load("ENV_MODEL/plain.pt")) 
>     x=model(state) 
>     return x 
>  
> state=torch.rand(1,10,13) 
> threads=[] 
> for i in range(5): 
>     p = multiprocessing.Process(target=run,args=(state,)) 
>     threads.append(p) 
> for i in threads: 
>     i.start() 
> for i in threads: 
>     i.join() 
> print("CHECKPOINT One") 
> run(state) 
> print("CHECKPOINT Two")
> threads=[] 
> for i in range(5): 
>     p = multiprocessing.Process(target=run,args=(state,)) 
>     threads.append(p) 
> for i in threads: 
>     i.start() 
> for i in threads: 
>     i.join()
> print("CHECKPOINT Three")
> ```
> 
> It got stuck after CHECKPOINT 2.

So, What I did:
- I had a master and slave agent protocol.
- In this Master will load this particular model for setting up the environment.
- Each slave will have a copy of the environment to gather, understand and contribute to another model which is learning to play in the environment.
- These Slave agents would be the child process and would train on copies of the environment with a different start position. 
- But if I don't load the model initially for the Master agent and just load in the slave agent, for each agent. It does not get stuck. 
- Once they gather the info, the master agent (in the parent process) will load the model.

**Comment by itzAmirali** [comment]
Did you guys find any solution? @GPaolo @BlackDChase 

I want to find the output of a model (eval mode) for a large dataset. What I did was use `multiprocessing` via `Process` in order to reduce the time. It seems that it is taking a lot more than if I do it with a `for`.  I have tried both defining one model externally and pass it to child processes, and also each child process defines its own model. None of them works!

Is there any way to solve it?

**Comment by miguelgondu** [comment]
> @GPaolo I meet the similar question. I use cpu, and get struck in function `load_state_dict`. Can you give some advises? Thanks.

I had a similar problem, and after an hour of debugging I realized that **calling another function that loaded the model** before the pool was causing the hang. So weird.

**Comment by shiqi-yang** [comment]
I was facing the similar problem but with Pool. I tried with multiprocessing.get_context("spawn").Pool, and it works, don't know why.

**Comment by sid548** [comment]
> @GPaolo I meet the similar question. I use cpu, and get struck in function `load_state_dict`. Can you give some advises? Thanks.

I am facing the similar issue. I have to make inferences from three different models, on the same dataset, using multiprocessing. The code is getting stuck in function 'load_state_dict'. Please provide me some inputs.

**Comment by xtzd** [comment]
> > @GPaolo I meet the similar question. I use cpu, and get struck in function `load_state_dict`. Can you give some advises? Thanks.
> 
> I had a similar problem, and after an hour of debugging I realized that **calling another function that loaded the model** before the pool was causing the hang. So weird.

I just meet the same issues (pytorch version 1.12.1+cu116). When I use load_state_dict before multiprocessing inference on CPU, it would stuck in the multiprocessing, while I donot use load_state_dict before multiprocessing inference, it works normally. It is so weird. I guess something is changed in the load_state_dict function.

**Comment by xtzd** [comment]
> @GPaolo I had a similar issue, then I did what @mruberry suggested: The first block of code works, the second block does not. (Both separate but similar files)
> 
> ```python
> import torch 
> from torch import multiprocessing,nn 
>  
> output_size = 13 
> input_dim = output_size 
> hidden_dim = 128 
> layer_dim = 1 
>  
> lstm = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
> torch.save(lstm.state_dict(),"ENV_MODEL/plain.pt") 
>  
> def run(state): 
>     model = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
>     model.load_state_dict(torch.load("ENV_MODEL/plain.pt")) 
>     x=model(state) 
>     return x 
>  
> state=torch.rand(1,10,13) 
> threads=[] 
> for i in range(5): 
>     p = multiprocessing.Process(target=run,args=(state,)) 
>     threads.append(p) 
> for i in threads: 
>     i.start() 
> for i in threads: 
>     i.join() 
> ```
> 
> This is done just to experiment, but it got stuck.
> 
> ```python
> import torch 
> from torch import multiprocessing,nn 
>  
> output_size = 13 
> input_dim = output_size 
> hidden_dim = 128 
> layer_dim = 1 
>  
> lstm = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
> torch.save(lstm.state_dict(),"ENV_MODEL/plain.pt") 
>  
> def run(state): 
>     model = nn.LSTM(input_size=input_dim, hidden_size=hidden_dim, num_layers=layer_dim, batch_first=True) 
>     model.load_state_dict(torch.load("ENV_MODEL/plain.pt")) 
>     x=model(state) 
>     return x 
>  
> state=torch.rand(1,10,13) 
> threads=[] 
> for i in range(5): 
>     p = multiprocessing.Process(target=run,args=(state,)) 
>     threads.append(p) 
> for i in threads: 
>     i.start() 
> for i in threads: 
>     i.join() 
> print("CHECKPOINT One") 
> run(state) 
> print("CHECKPOINT Two")
> threads=[] 
> for i in range(5): 
>     p = multiprocessing.Process(target=run,args=(state,)) 
>     threads.append(p) 
> for i in threads: 
>     i.start() 
> for i in threads: 
>     i.join()
> print("CHECKPOINT Three")
> ```
> 
> My actual code does involve more wrappers and class structure but later would be what o would do in a zest. Any suggestion for the latter? (It got stuck after CHECKPOINT 2.)

In this case, If you shrink the hidden_dim to a smaller number (below 90), It will not stuck either. But it is more weird.
